### ISSUE TITLE
Support for sqlite metadata for rpms in addition to XML

### ISSUE THREAD

**Issue by yavor-atanasov** [openSUSE/libsolv#92]
Hey guys,

Currently `libsolv` accepts only XML for the rpm metadata (filelists, primary etc). That metadata can be presented in the form of sqlite files as well. Would you be open for adding support for sqlite as well? Currently tools like `createrepo` which is predominantly used for generating repo metadata creates both XML and sqlite files. 

Let me give you our use case, could be useful. My team is responsible for services and tools that form a deployment pipeline to different platforms (cloud and on premise). Part of that ecosystem are a couple of services - one that bakes virtual machine images based on a set of repositories, packages and configuration, and another which is essentially a yum repository as a service - it allows our teams to create and maintain isolated repos. The repo service generates the yum metadata for a given repository on the fly when requested. Generating the metadata in the form of sqlite files is very fast for us as the data model behind the service mirrors closely the schema required for sqlite files (all we really do is copy required tables into sqlite, compress it and serve it).

We are migrating our bakery service to use `hawkey` (and thus `libsolv`) and `librepo` in order to speed up the baking process and do smarter things with pulling repo metadata, packages, installation etc. So now we've hit this problem where repositories served by our repo service cannot work with `libsolv` as it requires XML.

One of the reasons we do not generate XML files in the first place is because `yum` actually prefers sqlite and is faster using that. I realise the whole new `dnf` (the new fedora replacement for yum that uses libsolv) ecosystem don't want to just copy what `yum` does as there are many things wrong with it, however also `dnf` strives to be backwards compatible and I think this is one of those areas where it is not.

It would be great to get your thoughts on that. Thank you!

Yavor

**Comment by mlschroe** [comment]
First of all, this could be done either in libsolv or hawkey, so you may also want to ask the hawkey guys about their opinion.

I'm not opposed against adding sqlite support, as long as it is disabled by default (as I don't want to depend on a new library). Writing the converter should be pretty straghtforward.

Note that I'm on parential leave currently, so I'll only be able to work on this at the end of July.

**Comment by jsilhan** [comment]
Hi,
IIUC you want sqlite support just because of the better performance. DNF doesn't use it and probably will not use sqlite metadata files in the future so the format will be considered as obsoleted. I would rather suggest different solution to your use case. Frankly I don't like the idea of supporting sqlite format.

@mlschroe would it be possible to document the solv files format so createrepo can generate it? It would be IMO even more compressed and faster. DNF and PackageKit can benefit from it as well and eventually prefer this format.

@yavor-atanasov the metadata format will evolve and we will probably use different format than XML. Can you share your opinions on sqlite, XML, ... format it this BZ [1], please?

@dmach @Tojaj is there any progress of new metadata format. Have you considered the idea of solv files + binary diff?

/cc @rholy 

[1] https://bugzilla.redhat.com/show_bug.cgi?id=850896

**Comment by mlschroe** [comment]
Currently the solv files are meant as cache files, i.e. the reader code assumes that the content is somewhat sane. But while there are some rudimentary checks done it's possible to crash the reader with some bad data. So I would not recommend solv files as repo format because of security reasons.

It's not hard to make the reader completely verify the data, so in some future version things may be different.

**Comment by pclifford** [comment]
Hi, Yavor's away at the moment but I work with him.

We built our repository service to speed up the process of adding RPMs to a repository and installing them via yum.  The per-package metadata is extracted and persisted to a database as the RPM streams into the upload endpoint, and our createrepo equivalent just stitches together all the necessary rows to form the sqlite files at the point they're first requested.

We don't have strong attachments to any of the formats, but we're trying to minimise the overall time taken to both create the repository metadata and then to make use of it with either yum (our RHEL 6 and 7 clients) or dnf (Fedora, and presumably future RHEL 8 clients).

In one of our repositories of around 5000 (admittedly big) RPMs it takes about 4 seconds to fetch the source data from the database, 1 second to populate filelists.sqlite, 3.5 seconds to checksum/gzip/checksum it, and half a second for everything else.  The equivalent filelists.xml is 60-70% bigger so at least the checksum/gzip/checksum phase would take longer and then yum would just convert it into a smaller sqlite file anyway, so we don't generate XML metadata at the moment.

Given that dnf would similarly convert any XML we supplied into smaller solv files, of the available options we'd prefer to construct and supply solv files (or a hardened variant) directly if possible -- that would add the smallest overhead to the metadata generation step and get dnf started on useful work fastest.
